src/machine learning analysis.py

```diff
@@ -0,0 +1,155 @@
+        (OneHotEncoder(handle_unknown="ignore", sparse=False), categorical_feats),
+    )
+    
+    preprocessor.fit(X_train)
+    
+    # Get column names
+    columns = list(
+        preprocessor.named_transformers_["onehotencoder"].get_feature_names_out(
+            categorical_features
+        )
+    )
+    
+    # Pipeline including OneHotEncoder and LogisticRegression
+    pipe = make_pipeline(
+        OneHotEncoder(handle_unknown="ignore", sparse=False), 
+        LogisticRegression(max_iter=2000)
+    )
+    
+    # Scoring include accuracy, f1, recall, precision. 
+    results = {}
+    scoring = [
+        "accuracy",
+        "f1", 
+        "recall", 
+        "precision", 
+        "roc_auc", 
+        "average_precision"
+    ]
+    results["Logistic Regression"] = mean_std_cross_val_scores(pipe, X_train, y_train, scoring=scoring)
+    result_df = pd.DataFrame(results)
+    
+    # Plot confusion matrix
+    pipe.fit(X_train, y_train)
+    preds = pipe.predict(X_train)
+    conf_mat = confusion_matrix(y_train, preds)
+    conf_mat = pd.DataFrame(conf_mat)
+    
+    # Hyperparameter tuning using RandomSearchCV 
+    # Optimize using f1 as we have class imbalance.
+    param_grid = {
+        "logisticregression__C": loguniform(1e-3, 1e3),
+        "logisticregression__class_weight": [None, "balanced"],
+    }
+    random_search = RandomizedSearchCV(
+        pipe,
+        param_grid,
+        n_iter=50,
+        verbose=1,
+        n_jobs=-1,
+        random_state=123,
+        return_train_score=True,
+        scoring="f1"
+    )
+    random_search.fit(X_train, y_train)
+    print("Best hyperparameter values: ", random_search.best_params_)
+    print("Best score: %0.3f" % (random_search.best_score_))
+    
+    # Create output tables/images
+    output = opt["--output"]
+    save_df(result_df, "score_results")
+    save_df(conf_mat, "confusion matrix")
+    
+    # Storing optimized model
+    pickle.dump(model, open(f"{input}final_model.rds", "wb"))
+    
+# Function obtained from DSCI-571 lecture notes
+def mean_std_cross_val_scores(model, X_train, y_train, **kwargs):
+    """
+    Returns mean and std of cross validation
+
+    Parameters
+    ----------
+    model :
+        scikit-learn model
+    X_train : numpy array or pandas DataFrame
+        X in the training data
+    y_train :
+        y in the training data
+
+    Returns
+    ----------
+        pandas Series with mean scores from cross_validation
+    """
+
+    scores = cross_validate(model, X_train, y_train, **kwargs)
+
+    mean_scores = pd.DataFrame(scores).mean()
+    std_scores = pd.DataFrame(scores).std()
+    out_col = []
+
+    for i in range(len(mean_scores)):
+        out_col.append((f"%0.3f (+/- %0.3f)" % (mean_scores[i], std_scores[i])))
+
+    return pd.Series(data=out_col, index=mean_scores.index)
+    
+def save_df(df, name):
+    df.to_pickle(f"{output}{name}.rds")    
+    
+if __name__ == "__main__":
+    main() 
```

Comment:
Pass your docopt arguments in the function call here like: `main(opt["--input"])`.